plutus-core-interpreter/src/Language/PlutusCore/Interpreter/LMachine.hs

```diff
@@ -0,0 +1,284 @@
+-- | The L machine
+-- A lazy machine based on the L machine of Friedman et al. [Improving the Lazy Krivine Machine]
+-- For more details see the document in plutus/docs/fomega/lazy-machine
+-- The code here's closely based on the CEK machine implementation.
+
+-- The main difference from the CEK machine is that we have a heap
+-- containing closures containing evaluated/unevaluated function
+-- arguments, and environments mapping variable IDs to heap locations.
+-- To evaluate [M N], we save a closure C containing N and the current
+-- environment in a stack frame, then evaluate M to get a value
+-- (lambda/builtin) like \x.e and an environment env.  We retrieve the
+-- argument closure C from the stack, generate a new heap location l,
+-- bind C (unevaluated) to l in the heap, then evaluate e in
+-- env[x->l].  When we come across a use of x inside e, we look up l
+-- in the heap to get N (inside the closure C): if it's Evaluated then
+-- we continue immediately, using N instead of x. Otherwise we place
+-- an update marker on the stack and evaluate N to get a new closure
+-- C' containing a value V (and a possibly new environment): by
+-- this time we should be back at the update marker containing l, and
+-- we store Evaluated C back in the heap at l, then continue
+-- evaluating e.
+
+
+module Language.PlutusCore.Interpreter.LMachine
+    ( LMachineException
+    , EvaluationResultF (EvaluationSuccess, EvaluationFailure)
+    , EvaluationResult
+    , evaluateL
+    , runL
+    ) where
+
+import           Language.PlutusCore
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.View
+import           PlutusPrelude
+
+import           Control.Monad.Identity
+import           Data.IntMap                  (IntMap)
+import qualified Data.IntMap                  as IntMap
+
+type Plain f = f TyName Name ()
+
+-- Let's just throw exceptions for the time being: error handling is changeable at the moment
+-- FIXME: use the existing PLC exception infrastructure
+data LMachineException =
+    LMachineException String (Plain Term)
+  | LMachineStringException String
+    deriving (Show, Typeable)
+
+instance Exception LMachineException
+
+-- | A term together with an enviroment mapping free variables to heap locations
+data Closure = Closure
+    { _closureValue       :: Plain Term
+    , _closureEnvironment :: Environment
+    } deriving (Show)
+
+-- | L machine environments
+-- Each entry is a mapping from the 'Unique' representing a variable to a heap location
+newtype Environment = Environment (IntMap HeapLoc)
+    deriving (Show)
+
+-- | Heap location. Int should always be big enough: maxBound::Int is 2^63-1
```

Comment:
`Int` is platform dependent, it can be `Int32` as well. This is still big enough, but perhaps the comment should be changed.

Reply:
OK.